pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -149,206 +145,142 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	if errors.IsNotFound(err) {
 		// The resource no longer exists, in which case we stop processing.
 		c.Logger.Errorf("pipeline run %q in work queue no longer exists", key)
-		return nil
+		return controller.NewPermanentError(err)
 	} else if err != nil {
 		return err
 	}
 
 	// Don't modify the informer's copy.
 	pr := original.DeepCopy()
+
 	if !pr.HasStarted() {
+		// HasStarted is made false by InitializeConditions
 		pr.Status.InitializeConditions()
 		// In case node time was not synchronized, when controller has been scheduled to other nodes.
 		if pr.Status.StartTime.Sub(pr.CreationTimestamp.Time) < 0 {
 			c.Logger.Warnf("PipelineRun %s createTimestamp %s is after the pipelineRun started %s", pr.GetRunKey(), pr.CreationTimestamp, pr.Status.StartTime)
 			pr.Status.StartTime = &pr.CreationTimestamp
 		}
-		// start goroutine to track pipelinerun timeout only startTime is not set
+		// start goroutine to track pipelinerun timeout only if startTime is not set
 		go c.timeoutHandler.WaitPipelineRun(pr, pr.Status.StartTime)
-	} else {
-		pr.Status.InitializeConditions()
```

Comment:
why did we previously initializeconditions only if the pipelinerun had started and now we do the opposite (only if it has not started?)

Reply:
Previously we ran it both when pipeline had started and when not (L160 and L169).
I remember I thought a lot about this but it was sometime ago, now I need to look back into it

Reply:
`InitializeConditions` is idempotent so it is ok to run it everytime, but we don't have to.
It is convenient for some unit tests as the start time is then set, but I think it's better to set the start time explicitly in the unit tests where needed.